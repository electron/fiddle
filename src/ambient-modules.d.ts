declare module 'namor';
declare module 'shell-path';
