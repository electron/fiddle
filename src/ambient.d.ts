import * as MonacoType from 'monaco-editor';

import { App } from './renderer/app';

declare global {
  interface Window {
    ElectronFiddle: {
      app: App;
      appPaths: Record<string, string>;
      monaco: typeof MonacoType;
    };
  }
}
