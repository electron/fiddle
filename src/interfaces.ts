import { InstallState } from '@vertedinde/fiddle-core';

export type Files = Map<string, string>;

export type FileTransform = (files: Files) => Promise<Files>;

export enum VersionSource {
  remote = 'remote',
  local = 'local',
}

export enum GistActionType {
  publish = 'Publish',
  update = 'Update',
  delete = 'Delete',
}

export enum GistActionState {
  publishing = 'publishing',
  updating = 'updating',
  deleting = 'deleting',
  none = 'none',
}

export interface Version {
  version: string;
  name?: string;
  localPath?: string;
  node?: string;
}

export enum RunResult {
  SUCCESS = 'success', // exit code === 0
  FAILURE = 'failure', // ran, but exit code !== 0
  INVALID = 'invalid', // could not run
}

export interface RunnableVersion extends Version {
  state: InstallState;
  source: VersionSource;
  downloadProgress?: number;
}

export const enum ElectronReleaseChannel {
  stable = 'Stable',
  beta = 'Beta',
  nightly = 'Nightly',
}

export interface SetFiddleOptions {
  filePath?: string;
  templateName?: string;
  gistId?: string;
}

export interface SetUpMenuOptions {
  acceleratorsToBlock?: BlockableAccelerator[] | null;
  activeTemplate?: string | null;
}

export interface SetupRequest {
  fiddle?: SetFiddleOptions;
  version?: string;
  showChannels: ElectronReleaseChannel[];
  hideChannels: ElectronReleaseChannel[];
  useObsolete?: boolean;
}

export interface BisectRequest {
  setup: SetupRequest;
  goodVersion: string;
  badVersion: string;
}

export interface TestRequest {
  setup: SetupRequest;
}

export interface OutputEntry {
  text: string;
  timeString: string;
  isNotPre?: boolean;
}

export interface OutputOptions {
  bypassBuffer?: boolean;
  isNotPre?: boolean;
}

export interface GenericDialogOptions {
  type: GenericDialogType;
  ok: string;
  cancel?: string;
  wantsInput: boolean;
  defaultInput?: string;
  label: string | JSX.Element;
  placeholder?: string;
}

export interface Contributor {
  url: string;
  api: string;
  login: string;
  avatar: string;
  name: string;
  bio: string;
  location: string;
}

export interface Templates {
  [index: string]: string | Templates;
}

export const enum GenericDialogType {
  'confirm' = 'confirm',
  'warning' = 'warning',
  'success' = 'success',
}

export type EditorId = `${string}.${'js' | 'html' | 'css'}`;

export type EditorValues = Record<EditorId, string>;

// main.js gets special treatment: it is required as the entry point
// when we run fiddles or create a package.json to package fiddles.
export const MAIN_JS = 'main.js';

export const PACKAGE_NAME = 'package.json';

export type ArrowPosition = 'top' | 'left' | 'bottom' | 'right';

export const enum BlockableAccelerator {
  save = 'save',
  saveAs = 'saveAs',
}
