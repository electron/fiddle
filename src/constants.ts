export const SENTRY_DSN =
  'https://966a5b01ac8d4941b81e4ebd0ab4c991@sentry.io/1882540';
