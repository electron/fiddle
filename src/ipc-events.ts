export enum IpcEvents {
  OPEN_SETTINGS = 'OPEN_SETTINGS',
  LOAD_GIST_REQUEST = 'LOAD_GIST_REQUEST',
  LOAD_ELECTRON_EXAMPLE_REQUEST = 'LOAD_ELECTRON_EXAMPLE_REQUEST',
  FIDDLE_RUN = 'FIDDLE_RUN',
  FIDDLE_PACKAGE = 'FIDDLE_PACKAGE',
  FIDDLE_MAKE = 'FIDDLE_MAKE',
  MONACO_EXECUTE_COMMAND = 'MONACO_EXECUTE_COMMAND',
  MONACO_TOGGLE_OPTION = 'MONACO_TOGGLE_OPTION',
  FS_NEW_FIDDLE = 'FS_NEW_FIDDLE',
  FS_NEW_TEST = 'FS_NEW_TEST',
  FS_OPEN_FIDDLE = 'FS_OPEN_FIDDLE',
  FS_OPEN_TEMPLATE = 'FS_OPEN_TEMPLATE',
  FS_SAVE_FIDDLE = 'FS_SAVE_FIDDLE',
  FS_SAVE_FIDDLE_AS = 'FS_SAVE_FIDDLE_AS',
  FS_SAVE_FIDDLE_GIST = 'FS_SAVE_FIDDLE_GIST',
  FS_SAVE_FIDDLE_FORGE = 'FS_SAVE_FIDDLE_FORGE',
  FS_SAVE_FIDDLE_DIALOG = 'FS_SAVE_FIDDLE_DIALOG',
  FS_SAVE_FIDDLE_ERROR = 'FS_SAVE_FIDDLE_ERROR',
  SHOW_WARNING_DIALOG = 'SHOW_WARNING_DIALOG',
  SHOW_CONFIRMATION_DIALOG = 'SHOW_CONFIRMATION_DIALOG',
  SHOW_WELCOME_TOUR = 'SHOW_WELCOME_TOUR',
  CLEAR_CONSOLE = 'CLEAR_CONSOLE',
  LOAD_LOCAL_VERSION_FOLDER = 'LOAD_LOCAL_VERSION_FOLDER',
  SHOW_LOCAL_VERSION_FOLDER_DIALOG = 'SHOW_LOCAL_VERSION_FOLDER_DIALOG',
  BISECT_COMMANDS_TOGGLE = 'BISECT_COMMANDS_TOGGLE',
  SHOW_INACTIVE = 'SHOW_INACTIVE',
  BEFORE_QUIT = 'BEFORE_QUIT',
  CONFIRM_QUIT = 'CONFIRM_QUIT',
  GET_APP_PATHS = 'GET_APP_PATHS',
  SELECT_ALL_IN_EDITOR = 'SELECT_ALL_IN_EDITOR',
  BLOCK_ACCELERATORS = 'BLOCK_ACCELERATORS',
  SET_SHOW_ME_TEMPLATE = 'SET_SHOW_ME_TEMPLATE',
  CLICK_TITLEBAR_MAC = 'CLICK_TITLEBAR_MAC',
  TASK_BISECT = 'TASK_BISECT',
  TASK_TEST = 'TASK_TEST',
  TASK_DONE = 'TASK_DONE',
  OUTPUT_ENTRY = 'OUTPUT_ENTRY',
  RELOAD_WINDOW = 'RELOAD_WINDOW',
}

export const ipcMainEvents = [
  IpcEvents.FS_SAVE_FIDDLE_DIALOG,
  IpcEvents.FS_SAVE_FIDDLE,
  IpcEvents.SHOW_WARNING_DIALOG,
  IpcEvents.SHOW_CONFIRMATION_DIALOG,
  IpcEvents.SHOW_LOCAL_VERSION_FOLDER_DIALOG,
  IpcEvents.SHOW_INACTIVE,
  IpcEvents.CONFIRM_QUIT,
  IpcEvents.SET_SHOW_ME_TEMPLATE,
  IpcEvents.BLOCK_ACCELERATORS,
  IpcEvents.CLICK_TITLEBAR_MAC,
  IpcEvents.OUTPUT_ENTRY,
  IpcEvents.TASK_DONE,
  IpcEvents.RELOAD_WINDOW,
];

export const ipcRendererEvents = [
  IpcEvents.OPEN_SETTINGS,
  IpcEvents.LOAD_GIST_REQUEST,
  IpcEvents.LOAD_ELECTRON_EXAMPLE_REQUEST,
  IpcEvents.FIDDLE_RUN,
  IpcEvents.FIDDLE_PACKAGE,
  IpcEvents.FIDDLE_MAKE,
  IpcEvents.MONACO_EXECUTE_COMMAND,
  IpcEvents.MONACO_TOGGLE_OPTION,
  IpcEvents.FS_NEW_FIDDLE,
  IpcEvents.FS_NEW_TEST,
  IpcEvents.FS_OPEN_FIDDLE,
  IpcEvents.FS_OPEN_TEMPLATE,
  IpcEvents.FS_SAVE_FIDDLE,
  IpcEvents.FS_SAVE_FIDDLE_AS,
  IpcEvents.FS_SAVE_FIDDLE_FORGE,
  IpcEvents.FS_SAVE_FIDDLE_GIST,
  IpcEvents.SHOW_WELCOME_TOUR,
  IpcEvents.CLEAR_CONSOLE,
  IpcEvents.LOAD_LOCAL_VERSION_FOLDER,
  IpcEvents.BISECT_COMMANDS_TOGGLE,
  IpcEvents.BEFORE_QUIT,
  IpcEvents.GET_APP_PATHS,
  IpcEvents.SELECT_ALL_IN_EDITOR,
  IpcEvents.SET_SHOW_ME_TEMPLATE,
  IpcEvents.TASK_BISECT,
  IpcEvents.TASK_TEST,
];

export const WEBCONTENTS_READY_FOR_IPC_SIGNAL =
  'WEBCONTENTS_READY_FOR_IPC_SIGNAL';
