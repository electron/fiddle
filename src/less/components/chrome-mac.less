.darwin {
  .chrome {
    text-align: center;
    background-color: @background-1;
    border-bottom: @border;
    -webkit-app-region: drag;
    height: 24px;
    font-size: 1.4rem;
    line-height: 15px;

    small,
    p,
    span {
      user-select: none;
    }
  }

  .big_sur_or_later {
    height: 30px !important;
  }
}
