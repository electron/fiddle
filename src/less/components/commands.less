@import "../variables.less";

header {
  box-shadow: 0 0 0 1px rgba(16, 22, 26, 0.2), 0 0 0 rgba(16, 22, 26, 0), 0 1px 1px rgba(16, 22, 26, 0.4);
  background-color: @background-3;
  -webkit-app-region: drag;
}

.disabled-menu-tooltip {
  display: flex;
  width: 100%;

  .bp3-popover-target {
    width: 100%;
  }
}

.commands.is-mac {
  padding:
    10px
    calc(100vw - (env(titlebar-area-x, 0) + env(titlebar-area-width, 0)))
    10px
    env(titlebar-area-x, 0);
}

.commands {
  display: flex;
  justify-content: space-between;
  height: 50px;
  padding:
    10px;
  box-sizing: border-box;
  font-family: @fonts-common;

  button,
  input,
  select {
    -webkit-app-region: no-drag;
    margin-bottom: 0;
    white-space: nowrap;
  }

  select {
    width: 100px;
    margin-left: 5px;
  }

  fieldset {
    border: 0;
    padding: 0;
    margin: 0;
    min-width: 0;
  }

  & > div {
    display: flex;

    &:last-of-type {
      margin-right: 10px;

      .bp3-button-group {
        margin-left: 5px;
      }
    }
  }

  .bp3-control-group {
    margin-right: 0.5rem;
  }

  .address-bar {
    width: 420px;
    transition: width 0.4s;

    input,
    span {
      -webkit-app-region: no-drag;
    }

    &.empty {
      width: 300px;
    }

    &:focus-within {
      span,
      input {
        border-color: @foreground-2;
      }
    }
  }

  & > .title {
    align-self: stretch;
    flex-shrink: 0;

    flex-flow: column nowrap;
    justify-content: center;
    user-select: none;

    margin: 0 10px;
  }
}
