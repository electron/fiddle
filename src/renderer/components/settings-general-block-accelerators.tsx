import * as React from 'react';

import { Callout, Checkbox, FormGroup } from '@blueprintjs/core';
import { observer } from 'mobx-react';

import { BlockableAccelerator } from '../../interfaces';
import { AppState } from '../state';

interface BlockAcceleratorsSettingsProps {
  appState: AppState;
}

/**
 * Settings content to block keyboard accelerators.
 *
 * @class BlockAcceleratorsSettings
 * @extends {React.Component<BlockAcceleratorsSettingsProps>}
 */
export const BlockAcceleratorsSettings = observer(
  class BlockAcceleratorsSettings extends React.Component<BlockAcceleratorsSettingsProps> {
    constructor(props: BlockAcceleratorsSettingsProps) {
      super(props);

      this.handleBlockAcceleratorChange = this.handleBlockAcceleratorChange.bind(
        this,
      );
    }

    /**
     * Handles a change on whether a keyboard shortcut should be blocked
     * before fiddle is executed.
     *
     * @param {React.FormEvent<HTMLInputElement>} event
     */
    public handleBlockAcceleratorChange(
      event: React.FormEvent<HTMLInputElement>,
    ) {
      const { checked, value } = event.currentTarget;
      if (checked) {
        this.props.appState.addAcceleratorToBlock(
          value as BlockableAccelerator,
        );
      } else {
        this.props.appState.removeAcceleratorToBlock(
          value as BlockableAccelerator,
        );
      }
    }

    public render() {
      const { acceleratorsToBlock } = this.props.appState;

      const blockAcceleratorsInstructions = `
        Any keyboard shortcuts checked below will be disabled.`.trim();

      return (
        <div>
          <h4>Block Keyboard Shortcuts</h4>
          <Callout>
            <FormGroup>
              <p>{blockAcceleratorsInstructions}</p>
              <Checkbox
                checked={acceleratorsToBlock.includes(
                  BlockableAccelerator.save,
                )}
                label="Save"
                value={BlockableAccelerator.save}
                onChange={this.handleBlockAcceleratorChange}
              />
              <Checkbox
                checked={acceleratorsToBlock.includes(
                  BlockableAccelerator.saveAs,
                )}
                label="Save as"
                value={BlockableAccelerator.saveAs}
                onChange={this.handleBlockAcceleratorChange}
              />
            </FormGroup>
          </Callout>
        </div>
      );
    }
  },
);
