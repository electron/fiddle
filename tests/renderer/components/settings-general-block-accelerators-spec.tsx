import * as React from 'react';

import { shallow } from 'enzyme';

import { BlockableAccelerator } from '../../../src/interfaces';
import { BlockAcceleratorsSettings } from '../../../src/renderer/components/settings-general-block-accelerators';
import { StateMock } from '../../mocks/mocks';

describe('BlockAcceleratorsSettings component', () => {
  let store: StateMock;

  beforeEach(() => {
    ({ state: store } = (window as any).ElectronFiddle.app);
  });

  it('renders', () => {
    const wrapper = shallow(
      <BlockAcceleratorsSettings appState={store as any} />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  describe('handleBlockAcceleratorChange()', () => {
    it('handles a new selection', async () => {
      const wrapper = shallow(
        <BlockAcceleratorsSettings appState={store as any} />,
      );
      const instance = wrapper.instance() as any;

      await instance.handleBlockAcceleratorChange({
        currentTarget: { checked: false, value: BlockableAccelerator.save },
      });

      expect(store.removeAcceleratorToBlock).toHaveBeenCalledWith(
        BlockableAccelerator.save,
      );

      await instance.handleBlockAcceleratorChange({
        currentTarget: { checked: true, value: BlockableAccelerator.save },
      });

      expect(store.addAcceleratorToBlock).toHaveBeenCalledWith(
        BlockableAccelerator.save,
      );
    });
  });
});
