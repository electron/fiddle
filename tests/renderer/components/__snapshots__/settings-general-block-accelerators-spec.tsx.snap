// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlockAcceleratorsSettings component renders 1`] = `
<div>
  <h4>
    Block Keyboard Shortcuts
  </h4>
  <Blueprint3.Callout>
    <Blueprint3.FormGroup>
      <p>
        Any keyboard shortcuts checked below will be disabled.
      </p>
      <Blueprint3.Checkbox
        checked={false}
        label="Save"
        onChange={[Function]}
        value="save"
      />
      <Blueprint3.Checkbox
        checked={false}
        label="Save as"
        onChange={[Function]}
        value="saveAs"
      />
    </Blueprint3.FormGroup>
  </Blueprint3.Callout>
</div>
`;
